### DISCUSSION TITLE
Identity Provider using ConfiguredProvider but unable to inherit SAML UI fields

### DISCUSSION THREAD

**Discussion by nicolabeghin** [keycloak/keycloak#25862] [Q&A] [answered]
### Description

Hi everyone
I'm writing here since I'm lost with all the related `ConfiguredProvider` tickets/PRs, therefore pardon me if I'm missing something. 
* https://github.com/keycloak/keycloak/issues/15344
* https://github.com/keycloak/keycloak/discussions/11849
* https://github.com/keycloak/keycloak/issues/17958

I would like to know when [extending a SAML Identity Provider](https://github.com/italia/spid-keycloak-provider/pull/59/files#diff-2fdab77a8480f844b23b77f6d819904d0d39a07b605f040216f89fc7e9d67374) 

    extends AbstractIdentityProviderFactory<T> implements ConfiguredProvider

it's currently possible to:

# 1 - Usability - split fields in subsections
> Split `ProviderConfigProperty` in subsections `General Settings` vs `Advanced Settings` vs `Custom Settings`
<img width="1157" alt="image" src="https://github.com/keycloak/keycloak/assets/2743637/f77a841c-7cf2-4035-b3c7-93ffb52d87bb">

# 2- Inherit OOTB SAML IdP fields/UI
> Restore OOTB SAML fields (currently hardcoded in [DetailSettings.tsx#L397](https://github.com/keycloak/keycloak/blob/a228b6c7c9ec7a54ee91bb547b42cc4097ae38e2/js/apps/admin-ui/src/identity-providers/add/DetailSettings.tsx#L397))

    const isSAML = provider.providerId!.includes("saml");

<img width="1070" alt="image" src="https://github.com/italia/spid-keycloak-provider/assets/2743637/65c773ff-7e4d-4b92-9790-d53fb68c0147">

FYI here's my current PR where this is ongoing https://github.com/italia/spid-keycloak-provider/pull/59

thanks!
nicola

**Comment by Kagemusha-san** [comment]
I'm facing the same problem when trying to extend the OIDC Identity Provider. Based on your WIP changes I changed my `ProviderFactory` to implement `ConfiguredProvider` and added `getConfigProperties()` to my `ProviderConfig`. My provider is listed in Keycloak but only displays [GeneralSettings](https://github.com/keycloak/keycloak/blob/main/js/apps/admin-ui/src/identity-providers/add/GeneralSettings.tsx#L10) and my two custom fields.

The line you pointed out in [DetailSettings.tsx#L397](https://github.com/keycloak/keycloak/blob/a228b6c7c9ec7a54ee91bb547b42cc4097ae38e2/js/apps/admin-ui/src/identity-providers/add/DetailSettings.tsx#L397)

`const isSAML = provider.providerId!.includes("saml");`

suggests that it is enough to have `saml` or `oidc` in the `providerId`. So I appended `oidc`  to my `providerId`

`public static final String PROVIDER_ID = "myprovider-oidc";`

but it didn't work. Changing the `providerId` to `oidc` replaces the built in OIDC Provider with my own (at least the name suggests it) but does not display my custom fields.

I think [DetailSettings.tsx#L433-L445](https://github.com/keycloak/keycloak/blob/a228b6c7c9ec7a54ee91bb547b42cc4097ae38e2/js/apps/admin-ui/src/identity-providers/add/DetailSettings.tsx#L433-L445) explains this behavior. If the provider is neither `oidc` nor `saml`, `GeneralSettings` and `DynamicComponents` (aka custom fields), read from provider config, are displayed. If the provider is `oidc`, `OIDCGeneralSettings` will be displayed but `DynamicComponents` will not be added. Further down the file, OIDC and SAML components are hidden based on the value of `isOidc` and `isSaml`.

Since I never worked with React I might misunderstand some things but it seems that it's currently not possible to inherit OIDC/SAML UI fields AND add custom fields. I would be very happy if somebody teaches me better and points out how to achieve this 😉

**Reply by nicolabeghin** [reply]
> Changing the providerId to oidc replaces the built in OIDC Provider with my own (at least the name suggests it) but does not display my custom fields.

Just to confirm it's the same tentative I did for SAML and the same output I got for SAML 👍 

> it seems that it's currently not possible to inherit OIDC/SAML UI fields AND add custom fields

Totally agree with your conclusion, IMHO some rework is required on Keycloak-side (not possible in custom provider) to be able to inherit OIDC/SAML UI fields **AND** add custom fields this

**Reply by nicolabeghin** [reply]
Some good news from upstream commits https://github.com/keycloak/keycloak/commit/67912209aedd3c3b861e5bc301e20e86e9184403 (add) and https://github.com/keycloak/keycloak/commit/67912209aedd3c3b861e5bc301e20e86e9184403 (view) allowing OOTB SAML/OIDC fields to be shown in addition to custom ones, exactly as per our case - PR https://github.com/keycloak/keycloak/issues/25666 and https://github.com/keycloak/keycloak/discussions/25862

I will also try to submit a way to generalize SAML/OIDC detection in place of hardcoded "saml"/"oidc" text pattern then we should be all good to go.

**Reply by nosnilmot** [reply]
> I will also try to submit a way to generalize SAML/OIDC detection in place of hardcoded "saml"/"oidc" text pattern then we should be all good to go.

Should I assume, with this discussion now 'closed', that you are no longer planning to do this?

**Reply by nicolabeghin** [reply]
> > I will also try to submit a way to generalize SAML/OIDC detection in place of hardcoded "saml"/"oidc" text pattern then we should be all good to go.
> 
> Should I assume, with this discussion now 'closed', that you are no longer planning to do this?

I confirm, in the end I switched the provider ID to conform with hardcoded Keycloak 24 logic (it's a simple change in code and just requires an SQL query in the keycloak db)

**Reply by nosnilmot** [reply]
Thanks for confirming.

It's trivial to change ... if you don't have clients referencing the existing provider ID in auth/token URLs 😞  and thanks to https://github.com/keycloak/keycloak/issues/24710 I can't configure a duplicate provider with new ID and migrate clients over to it gradually (of course I can duplicate the whole provider code, but ...)

**Reply by nosnilmot** [reply]
Oh! the same hardcoded provider ID matching (saml/oidc) is used to determine if 'alias' can be set when adding an IDP, and changing provider ID does not change alias for existing IDPs, so actually changing provider ID should work fine for me too.

**Comment by nicolabeghin** [comment, marked as answer]
Closing as fixed with Keycloak 24.x
